### PR TITLE
[16.0][MIG] purchase_order_secondary_unit

### PR DESCRIPTION
**Pull request by dalonsod**
Supersedes #2068

@SirAionTech tried to make every change suggested at that PR (removed unnecessary changes and fix splitted in a standalone commit). One test fails (at least locally), still need work.

cc @acsonefho @anddago78

### CODE REVIEW

**Comment by SirAionTech** [comment]
:+1: great, let me know when CI works.
In the meanwhile, you can set the PR as draft (https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request#converting-a-pull-request-to-a-draft) since it still needs work.

**Comment by dalonsod** [comment]
@SirAionTech done!!!

File: purchase_order_secondary_unit/models/purchase_order.py
Comment (SirAionTech) [inline purchase_order_secondary_unit/models/purchase_order.py]: There is a big warning in the tests due to this and similar changes
> 2024-04-05 11:19:00,359 446 WARNING odoo py.warnings: /opt/odoo/odoo/fields.py:461: UserWarning: precompute attribute doesn't make any sense on non computed field purchase.order.line.product_qty

(from https://github.com/OCA/purchase-workflow/actions/runs/8568957322/job/23483847579?pr=2216#step:8:35)
could you please clarify why this is needed? Or maybe it's a false positive?

It is raised in https://github.com/odoo/odoo/blob/38ec36e9b5dde0c02a4ff47714b1056e0db76812/odoo/fields.py#L461 and it also unsets `precompute` parameter for the field; maybe we simply should assign the `compute` and `store` attributes to make the core happy, what do you think?

File: purchase_order_secondary_unit/tests/test_purchase_order_secondary_unit.py
Comment (SirAionTech) [inline purchase_order_secondary_unit/tests/test_purchase_order_secondary_unit.py]: I am concerned about this change: the test was expecting `line.secondary_uom_qty == 3.5`, now it happens that `line.secondary_uom_qty == 3500`.
Usually when this happens it means that there might be something to be fixed in the code, or some cache value to be reset in the test, but the test's assertions shouldn't be changed in a migration.

Or maybe are you saying that the test was asserting something wrong already in `15.0`?

File: purchase_order_secondary_unit/tests/test_purchase_order_secondary_unit.py
Comment (SirAionTech) [inline purchase_order_secondary_unit/tests/test_purchase_order_secondary_unit.py]: If this reordering is not strictly needed, I would leave it as it was, so that in the migration we only have what is needed for the migration itself.

File: purchase_order_secondary_unit/models/purchase_order.py
Comment (SirAionTech) [inline purchase_order_secondary_unit/models/purchase_order.py]: Could you please clarify what the bug was?

File: purchase_order_secondary_unit/views/purchase_order_views.xml
Comment (SirAionTech) [inline purchase_order_secondary_unit/views/purchase_order_views.xml:14]: Why isn't the same `t` tag needed in the other `xpath` below?

**Review by SirAionTech — CHANGES_REQUESTED** [review]
Thanks for the PR!
I only looked at the code.

It would be nice if all the migration changes that do not have to be backported could go in the same **[MIG]** commit, maybe you could set yourself as co-author of the **[MIG]** commit, if the original author @yajo agrees?

Could you please also squash the bots commits as explained in https://github.com/OCA/maintainer-tools/wiki/Merge-commits-in-pull-requests#mergesquash-the-commits-generated-by-bots-or-weblate?

**Comment by dalonsod** [comment]
hello @SirAionTech thanks for the comments. Most of them are related to changes made by other contributors, I've only superseded them, prorizing commits authorship and thus, last changes made by other people, so I'm not able to answer some of that questions you've made. On my own, I solved your last questions about separating commits and fixed a pending test, supposing other changes in core were right, indeed (maybe I did a bad choice).

I could rollback every change that seems to be unnecessary or inadequate (or simply start from scratch), but I'd wait for @acsonefho and @anddago78 feedback

File: purchase_order_secondary_unit/tests/test_purchase_order_secondary_unit.py
Comment (dalonsod) [reply purchase_order_secondary_unit/tests/test_purchase_order_secondary_unit.py]: Ok, it's my fault, I'll fix it

File: purchase_order_secondary_unit/models/purchase_order.py
Comment (acsonefho) [reply purchase_order_secondary_unit/models/purchase_order.py]: I already have this issue. To avoid the warning, we have to duplicate some attributes (like already discussed here: https://github.com/OCA/purchase-workflow/pull/2068/files#r1415171629 ).
I agree with the comment: we shouldn't have to duplicate attribute. So we have to do a choice between duplicates and the warning.

**Comment by yajo** [comment]
I think it's good to respect others' authorship. But if you need to squash some things because some commits won't have sense in the final consolidated history, at least [tag them as co-authored](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors).

File: purchase_order_secondary_unit/views/purchase_order_views.xml
Comment (dalonsod) [reply purchase_order_secondary_unit/views/purchase_order_views.xml:14]: Added group coverage for missing fields (`groups` attribute instead of `<t>`, that cannot be used under a `<tree>` view)

File: purchase_order_secondary_unit/models/purchase_order.py
Comment (dalonsod) [reply purchase_order_secondary_unit/models/purchase_order.py]: This is explained by original author in commit description

> `Fix a bug when calling super's `_compute_product_qty()` was skipping the compute helper from the mixin.`

File: purchase_order_secondary_unit/models/purchase_order.py
Comment (dalonsod) [reply purchase_order_secondary_unit/models/purchase_order.py]: Restored duplicated attributes in order to avoid warnings

File: purchase_order_secondary_unit/tests/test_purchase_order_secondary_unit.py
Comment (dalonsod) [reply purchase_order_secondary_unit/tests/test_purchase_order_secondary_unit.py]: Fixed

**Comment by dalonsod** [comment]
> It would be nice if all the migration changes that do not have to be backported could go in the same **[MIG]** commit, maybe you could set yourself as co-author of the **[MIG]** commit, if the original author @yajo agrees?
> 
> Could you please also squash the bots commits as explained in https://github.com/OCA/maintainer-tools/wiki/Merge-commits-in-pull-requests#mergesquash-the-commits-generated-by-bots-or-weblate?

I've preserved changes in separated commits because one of them was marked as fix by original author, so it can be backported.

And merged bots commits as required

File: purchase_order_secondary_unit/tests/test_purchase_order_secondary_unit.py
Comment (dalonsod) [reply purchase_order_secondary_unit/tests/test_purchase_order_secondary_unit.py]: I'm not sure why these changes were made, maybe @yajo could help us, as original change (at least partially) was made by him, if I'm not wrong

**Comment by dalonsod** [comment]
Hello, @SirAionTech could you review again? Thanks!

File: purchase_order_secondary_unit/models/purchase_order.py
Comment (SirAionTech) [reply purchase_order_secondary_unit/models/purchase_order.py]: I had read that already, but saying that the bug was "the helper was not called" does not help me understand what the bug was.

Why do we need to call the helper? What is the error or issue happening without this change?

File: purchase_order_secondary_unit/views/purchase_order_views.xml
Comment (SirAionTech) [reply purchase_order_secondary_unit/views/purchase_order_views.xml:14]: Thanks

File: purchase_order_secondary_unit/models/purchase_order.py
Comment (SirAionTech) [reply purchase_order_secondary_unit/models/purchase_order.py]: Thanks

File: purchase_order_secondary_unit/tests/test_purchase_order_secondary_unit.py
Comment (SirAionTech) [reply purchase_order_secondary_unit/tests/test_purchase_order_secondary_unit.py]: Thanks

**Review by SirAionTech — CHANGES_REQUESTED** [review]
Thanks for squashing the commits! I have noticed that https://github.com/OCA/purchase-workflow/commit/4b1b2f8d572d7bb2e76d013e3bc2b03648866c56 is missing, maybe it has been squashed by mistake, could you please check?

Also some threads from previous review https://github.com/OCA/purchase-workflow/pull/2216#pullrequestreview-1988333356 still have to be resolved.

**Comment by pilarvargas-tecnativa** [comment]
please include https://github.com/OCA/purchase-workflow/pull/2311

**Comment by sergio-teruel** [comment]
@dalonsod Can you add this commit from 15 7b728cd4ba9b4b2ec58cc545fdd48e99b8812974?

Do you have planned finished the migration work?

Thanks

**Comment by dalonsod** [comment]
> please include #2311

@pilarvargas-tecnativa added

> Can you add this commit from 15 https://github.com/OCA/purchase-workflow/commit/7b728cd4ba9b4b2ec58cc545fdd48e99b8812974?

@sergio-teruel added

**Comment by dalonsod** [comment]
> Do you have planned finished the migration work?

@sergio-teruel this PR has been a little headache for me, because I've superseded a PR that had been superseded another one as well. I've tried to preserve some improvements/contributions added by previous authors, some of them technically complex for my current knowledge of recent Odoo versions, or difficult to understand. Most of pending tasks are finished IMO, but some of them are still pending.

Not sure if I can resume this in the next weeks, so feel free to supersede it, if you're ready for it...

**Review by sergio-teruel — APPROVED** [review]
Functional review OK
code review!

**Comment by SirAionTech** [comment]
> /ocabot migration purchase_order_secondary_unit
> /ocabot merge nobump

@pedrobaeza some things I mentioned in https://github.com/OCA/purchase-workflow/pull/2216#pullrequestreview-2085322687 are still missing, for instance commit https://github.com/OCA/purchase-workflow/commit/4b1b2f8d572d7bb2e76d013e3bc2b03648866c56 has not been ported: it has probably been squashed by mistake in some other commit.

**Comment by pedrobaeza** [comment]
Sorry, didn't know. Can any of you do the corresponding followup PR?

**Comment by SirAionTech** [comment]
> Sorry, didn't know. Can any of you do the corresponding followup PR?

For what I see, the open comments cannot be fixed afterwards.

I wouldn't know how to do a PR to restore a commit that has been squashed in some other commit, even if I added an empty commit (just to mention that the commit's author helped with the translations) the history would be wrong.

Another unresolved comment is for understanding why a change has been made, for instance:
> Why do we need to call the helper? What is the error or issue happening without this change?

(from https://github.com/OCA/purchase-workflow/pull/2216#discussion_r1618833459).
I couldn't do a PR to fix that either.

I was just waiting for some answers and being asked to update the review.
Maybe the comments were not important enough to block the migration, but an answer would have been appreciated anyway.

Just please have a better look at
image
before merging.

**Comment by pedrobaeza** [comment]
Well, the usual flow for blocking a PR is to put "Changes requested". Sorry anyway for not looking with enough care.

**Comment by SirAionTech** [comment]
> Well, the usual flow for blocking a PR is to put "Changes requested".

That is what I did in https://github.com/OCA/purchase-workflow/pull/2216#pullrequestreview-2085322687
image
and that results in 
image
(the tooltip is shown on hover)

Please let me know if I had to do something different: if I'm reviewing things wrong I'd like to know before creating more confusion in other pull requests.

> Sorry anyway for not looking with enough care.

Appreciate it, thanks :)

**Comment by pedrobaeza** [comment]
No, you didn't. You pressed on "Comment", not on "Request changes". You did code suggestions, but that's something different.

imagen

**Pull request merged**
